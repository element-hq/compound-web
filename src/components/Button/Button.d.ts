import React, { PropsWithChildren } from "react";
type ButtonProps<C extends React.ElementType> = {
    as?: C;
    kind?: "primary" | "secondary" | "tertiary" | "destructive";
    size?: "sm" | "lg";
    className?: string;
} & React.ComponentPropsWithoutRef<C>;
export declare const Button: <C extends React.ElementType<any> = "button">({ as, kind, size, children, className, ...props }: React.PropsWithChildren<ButtonProps<C>>) => React.JSX.Element;
export {};
