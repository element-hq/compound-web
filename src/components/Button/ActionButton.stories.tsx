/*
Copyright 2023 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React from "react";
import { type Meta, type StoryObj } from "@storybook/react-vite";
import { fn } from "storybook/test";

import * as icons from "@vector-im/compound-design-tokens/assets/web/icons";

import { Button as ButtonComponent } from "./Button";
import { Tooltip } from "../Tooltip/Tooltip";
import { TooltipProvider } from "../Tooltip/TooltipProvider";

const Template: React.FC<
  { label: string; as: "button" | "a" } & React.ComponentProps<
    typeof ButtonComponent
  >
> = ({ label, ...args }) => (
  <TooltipProvider>
    <Tooltip label={label}>
      <ButtonComponent iconOnly {...args} />
    </Tooltip>
  </TooltipProvider>
);

const meta = {
  title: "Button/Actions",
  component: Template,
  tags: ["autodocs"],
  argTypes: {
    label: {
      type: "string",
    },
    size: {
      options: ["sm", "lg"],
      control: { type: "inline-radio" },
    },
    kind: {
      options: ["primary", "secondary", "tertiary"],
      control: { type: "inline-radio" },
    },
    destructive: { type: "boolean" },
    disabled: {
      type: "boolean",
    },
    as: {
      options: ["a", "button"],
      control: { type: "inline-radio" },
    },
    Icon: {
      options: Object.keys(icons),
      mapping: icons,
      control: { type: "select" },
    },
    onClick: { action: "onClick" },
  },
  args: {
    label: "Action",
    size: "lg",
    as: "button",
    destructive: false,
    disabled: false,
    onClick: fn(),
  },
  parameters: {
    controls: {
      exclude: ["children"],
    },
  },
} satisfies Meta<typeof Template>;
export default meta;

type Story = StoryObj<typeof meta>;

export const Settings: Story = {
  args: {
    label: "Settings",
    kind: "secondary",
    Icon: icons.SettingsSolidIcon,
  },
};

export const ShareLink: Story = {
  args: {
    label: "Copy link",
    kind: "secondary",
    Icon: icons.LinkIcon,
  },
};

export const VideoOn: Story = {
  args: {
    label: "Turn video off",
    kind: "secondary",
    Icon: icons.VideoCallSolidIcon,
  },
};

export const VideoOff: Story = {
  args: {
    label: "Turn video on",
    kind: "primary",
    Icon: icons.VideoCallOffSolidIcon,
  },
};

export const MicOn: Story = {
  args: {
    label: "Turn mic off",
    kind: "secondary",
    Icon: icons.MicOnSolidIcon,
  },
};

export const MicOff: Story = {
  args: {
    label: "Turn mic on",
    kind: "primary",
    Icon: icons.MicOffSolidIcon,
  },
};

export const ShareScreen: Story = {
  args: {
    label: "Share screen",
    kind: "secondary",
    Icon: icons.ShareScreenSolidIcon,
  },
};

export const SharingScreen: Story = {
  args: {
    label: "Stop sharing screen",
    kind: "primary",
    Icon: icons.ShareScreenSolidIcon,
  },
};

export const Fullscreen: Story = {
  args: {
    label: "Full screen",
    kind: "secondary",
    Icon: icons.ExpandIcon,
  },
};

export const EndCall: Story = {
  args: {
    label: "End call",
    kind: "primary",
    destructive: true,
    Icon: icons.EndCallIcon,
  },
};

export const Invite: Story = {
  args: {
    label: "Invite",
    kind: "primary",
    size: "sm",
    Icon: icons.UserAddIcon,
  },
};

export const Edit: Story = {
  args: {
    label: "Edit",
    kind: "primary",
    size: "sm",
    Icon: icons.EditIcon,
  },
};

export const Search: Story = {
  args: {
    label: "Search",
    kind: "secondary",
    size: "sm",
    Icon: icons.SearchIcon,
  },
};

export const Save: Story = {
  args: {
    label: "Save",
    kind: "primary",
    size: "sm",
    Icon: icons.CheckIcon,
  },
};

export const SaveDisabled: Story = {
  args: {
    label: "Save",
    kind: "primary",
    size: "sm",
    Icon: icons.CheckIcon,
    disabled: true,
  },
};

export const Cancel: Story = {
  args: {
    label: "Cancel",
    kind: "secondary",
    size: "sm",
    Icon: icons.CloseIcon,
  },
};

export const Close: Story = {
  args: {
    label: "Close",
    kind: "secondary",
    size: "sm",
    Icon: icons.CloseIcon,
  },
};
