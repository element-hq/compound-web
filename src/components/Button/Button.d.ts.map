{"version":3,"file":"Button.d.ts","sourceRoot":"","sources":["../../../../src/components/Button/Button.tsx"],"names":[],"mappings":"AAiBA,OAAO,KAAK,EAAE,EAAE,iBAAiB,EAAE,MAAM,OAAO,CAAC;AAGjD,KAAK,WAAW,CAAC,CAAC,SAAS,KAAK,CAAC,WAAW,IAAI;IAC9C,EAAE,CAAC,EAAE,CAAC,CAAC;IACP,IAAI,CAAC,EAAE,SAAS,GAAG,WAAW,GAAG,UAAU,GAAG,aAAa,CAAC;IAC5D,IAAI,CAAC,EAAE,IAAI,GAAG,IAAI,CAAC;IACnB,SAAS,CAAC,EAAE,MAAM,CAAC;CACpB,GAAG,KAAK,CAAC,wBAAwB,CAAC,CAAC,CAAC,CAAC;AAEtC,eAAO,MAAM,MAAM,+IAOoB,SAAS,QAkB/C,CAAC"}