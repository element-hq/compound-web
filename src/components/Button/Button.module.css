/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.button {
  border-radius: var(--cpd-radius-pill-effect);
  cursor: pointer;
  appearance: none;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: var(--cpd-space-2x);
  box-sizing: border-box;
  font: var(--cpd-font-body-md-semibold);
}

.button[aria-disabled="true"] {
  cursor: not-allowed;
  pointer-events: all !important;
}

/**
 * SIZES
 */

.button[data-size="lg"] {
  padding-block: var(--cpd-space-2x);
  padding-inline: var(--cpd-space-8x);
  min-block-size: var(--cpd-space-12x);
}

.button[data-size="lg"].has-icon {
  padding-inline-start: var(--cpd-space-7x);
}

.button[data-size="sm"] {
  padding-block: var(--cpd-space-1x);
  padding-inline: var(--cpd-space-6x);
  min-block-size: var(--cpd-space-9x);
}

.button[data-size="sm"].has-icon {
  padding-inline-start: var(--cpd-space-5x);
}

/**
 * KINDS
 */

.button[data-kind="primary"] {
  color: var(--cpd-color-text-on-solid-primary);
  background: var(--cpd-color-bg-action-primary-rest);
  border-width: 0;
}

@media (hover) {
  .button[data-kind="primary"]:hover {
    background: var(--cpd-color-bg-action-primary-hovered);
  }
}

.button[data-kind="primary"]:active,
.button[data-kind="primary"][aria-expanded="true"] {
  background: var(--cpd-color-bg-action-primary-pressed);
}

.button[data-kind="primary"][aria-disabled="true"] {
  color: var(--cpd-color-text-disabled);
  background: var(--cpd-color-bg-subtle-primary);
}

.button[data-kind="secondary"] {
  border: 1px solid var(--cpd-color-border-interactive-primary);
  color: var(--cpd-color-text-primary);
  background: var(--cpd-color-bg-canvas-default);
}

@media (hover) {
  .button[data-kind="secondary"]:hover {
    border-color: var(--cpd-color-border-interactive-hovered);
    background: var(--cpd-color-bg-subtle-secondary);
  }
}

.button[data-kind="secondary"]:active,
.button[data-kind="secondary"][aria-expanded="true"] {
  border-color: var(--cpd-color-border-interactive-hovered);
  background: var(--cpd-color-bg-subtle-primary);
}

.button[data-kind="secondary"][aria-disabled="true"] {
  border-color: var(--cpd-color-border-interactive-secondary);
  color: var(--cpd-color-text-disabled);
  background: var(--cpd-color-bg-subtle-secondary);
}

.button[data-kind="tertiary"] {
  border: none;
  color: var(--cpd-color-text-primary);
  text-decoration: underline;
  background: transparent;
}

@media (hover) {
  .button[data-kind="tertiary"]:hover {
    background: var(--cpd-color-bg-subtle-secondary);
  }
}

.button[data-kind="tertiary"]:active,
.button[data-kind="tertiary"][aria-expanded="true"] {
  background: var(--cpd-color-bg-subtle-primary);
}

.button[data-kind="tertiary"][aria-disabled="true"] {
  color: var(--cpd-color-text-disabled);
}

.button[data-kind="destructive"] {
  border: 1px solid var(--cpd-color-border-critical-primary);
  color: var(--cpd-color-text-critical-primary);
  background: var(--cpd-color-bg-canvas-default);
}

@media (hover) {
  .button[data-kind="destructive"]:hover {
    border-color: var(--cpd-color-border-critical-hovered);
    background: var(--cpd-color-bg-critical-subtle);
  }
}

.button[data-kind="destructive"]:active,
.button[data-kind="destructive"][aria-expanded="true"] {
  border-color: var(--cpd-color-border-critical-hovered);
  background: var(--cpd-color-bg-critical-subtle-hovered);
}

.button[data-kind="destructive"][aria-disabled="true"] {
  border-color: var(--cpd-color-border-interactive-secondary);
  color: var(--cpd-color-text-disabled);
  background: var(--cpd-color-bg-subtle-secondary);
}
