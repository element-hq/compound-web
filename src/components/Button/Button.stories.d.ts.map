{"version":3,"file":"Button.stories.d.ts","sourceRoot":"","sources":["../../../../src/components/Button/Button.stories.tsx"],"names":[],"mappings":"AAgBA,OAAO,KAAK,MAAM,OAAO,CAAC;;;;;;;;;AAK1B,wBAUkC;AAMlC,eAAO,MAAM,OAAO;;;;;;;EAAoB,CAAC;AAYzC,eAAO,MAAM,SAAS;;;;;;;EAAoB,CAAC;AAY3C,eAAO,MAAM,QAAQ;;;;;;;EAAoB,CAAC;AAK1C,eAAO,MAAM,WAAW;;;;;;;EAAoB,CAAC"}