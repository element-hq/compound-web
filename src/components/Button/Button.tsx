/*
Copyright 2023  New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import classNames from "classnames";
import React, {
  ComponentType,
  PropsWithChildren,
  forwardRef,
  ForwardedRef,
  Ref,
} from "react";
import styles from "./Button.module.css";
import { UnstyledButton, UnstyledButtonPropsFor } from "./UnstyledButton";

interface ButtonComponent {
  // With the explicit `as` prop
  <C extends React.ElementType>(
    props: { as: C } & ButtonPropsFor<C>,
  ): React.ReactElement;
  // Without the explicit `as` prop, defaulting to a <button>
  (props: ButtonPropsFor<"button">): React.ReactElement;
}

type ButtonOwnProps = PropsWithChildren<{
  /**
   * The type of button.
   */
  kind?: "primary" | "secondary" | "tertiary" | "destructive"; // TODO: Refine the naming
  /**
   * The t-shirt size of the button.
   */
  size?: "sm" | "lg";
  /**
   * An icon to display within the button.
   */
  Icon?: ComponentType<React.SVGAttributes<SVGElement>>;
}>;

type ButtonPropsFor<C extends React.ElementType> = ButtonOwnProps &
  UnstyledButtonPropsFor<C>;

/**
 * A button component that can be transformed into a link, but keep the button
 * styling using the `as` property.
 */
export const Button = forwardRef(function Button<
  C extends React.ElementType = "button",
>(
  {
    as,
    kind = "primary",
    size = "lg",
    children,
    className,
    Icon,
    disabled,
    ...props
  }: ButtonPropsFor<C> & { as?: C },
  ref: ForwardedRef<C>,
): React.ReactElement {
  const classes = classNames(styles.button, className, {
    [styles["has-icon"]]: Icon,
  });

  return (
    <UnstyledButton
      {...props}
      as={as || ("button" as const)}
      ref={ref as Ref<C>}
      className={classes}
      data-size={size}
      data-kind={kind}
      tabIndex={0}
      disabled={disabled}
    >
      {Icon && (
        <Icon
          width={20}
          height={20}
          className={styles.icon}
          aria-hidden={true}
        />
      )}
      {children}
    </UnstyledButton>
  );
}) as ButtonComponent;
