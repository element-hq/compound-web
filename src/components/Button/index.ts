/*
Copyright 2023 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

export { Button } from "./Button";
export { IconButton } from "./IconButton/IconButton";
