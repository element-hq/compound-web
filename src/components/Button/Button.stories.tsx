/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { Meta } from "@storybook/react";

import { Button as ButtonComponent } from "./Button";
import VisibilityOnIcon from "@vector-im/compound-design-tokens/icons/visibility-on.svg";

export default {
  title: "Button",
  component: ButtonComponent,
  tags: ["autodocs"],
  parameters: {
    actions: { argTypesRegex: "^on.*" },
  },
  argTypes: {
    size: {
      options: ["sm", "lg"],
      control: { type: "inline-radio" },
    },
    kind: {
      options: ["primary", "secondary", "tertiary", "destructive"],
      control: { type: "inline-radio" },
    },
    disabled: {
      type: "boolean",
    },
    as: {
      options: ["a", "button"],
      control: { type: "inline-radio" },
    },
    Icon: {
      options: [undefined, "VisibilityOnIcon"],
      mapping: {
        VisibilityOnIcon: VisibilityOnIcon,
        undefined: undefined,
      },
      control: { type: "inline-radio" },
    },
    onClick: { action: "onClick" },
  },
  args: {
    size: "lg",
    as: undefined,
    disabled: false,
    children: "Click me!",
  },
} as Meta<typeof ButtonComponent>;

export const Default = {
  args: {
    // test component defaults
    kind: undefined,
    size: undefined,
  },
};

export const Small = {
  args: {
    // test component defaults
    kind: undefined,
    size: "sm",
  },
};

export const Primary = {
  args: {
    kind: "primary",
  },
};

export const Secondary = {
  args: {
    kind: "secondary",
  },
};

export const Tertiary = {
  args: {
    kind: "tertiary",
  },
};

export const Destructive = {
  args: {
    kind: "destructive",
  },
};

export const WithIcon = {
  args: {
    ...Primary.args,
    Icon: VisibilityOnIcon,
  },
};

export const SmallWithIcon = {
  args: {
    ...Primary.args,
    size: "sm",
    Icon: VisibilityOnIcon,
  },
};

export const Disabled = {
  args: {
    ...Primary.args,
    disabled: true,
  },
};

export const Link = {
  args: {
    ...Primary.args,
    as: "a",
    href: "https://example.org",
  },
};

export const LinkDisabled = {
  args: {
    ...Link.args,
    disabled: true,
  },
};
