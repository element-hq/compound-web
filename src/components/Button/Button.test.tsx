/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { vi, describe, it, expect } from "vitest";
import React from "react";
import { fireEvent, screen, render } from "@testing-library/react";

import { Button } from "./Button";

import * as stories from "./Button.stories";
import { composeStories } from "@storybook/react";

const {
  Default,
  Small,
  Primary,
  Secondary,
  Tertiary,
  Destructive,
  WithIcon,
  SmallWithIcon,
  Disabled,
  Link,
  LinkDisabled,
} = composeStories(stories);

describe("Button", () => {
  it("renders a Default button", () => {
    const { container } = render(<Default />);
    expect(container).toMatchSnapshot();
  });
  it("renders a Small button", () => {
    const { container } = render(<Small />);
    expect(container).toMatchSnapshot();
  });
  it("renders a Primary button", () => {
    const { container } = render(<Primary />);
    expect(container).toMatchSnapshot();
  });
  it("renders a Secondary button", () => {
    const { container } = render(<Secondary />);
    expect(container).toMatchSnapshot();
  });
  it("renders a Tertiary button", () => {
    const { container } = render(<Tertiary />);
    expect(container).toMatchSnapshot();
  });
  it("renders a Destructive button", () => {
    const { container } = render(<Destructive />);
    expect(container).toMatchSnapshot();
  });
  it("renders a WithIcon button", () => {
    const { container } = render(<WithIcon />);
    expect(container).toMatchSnapshot();
  });
  it("renders a SmallWithIcon button", () => {
    const { container } = render(<SmallWithIcon />);
    expect(container).toMatchSnapshot();
  });
  it("renders a Disabled button", () => {
    const { container } = render(<Disabled />);
    expect(container).toMatchSnapshot();
  });
  it("renders a Link button", () => {
    const { container } = render(<Link />);
    expect(container).toMatchSnapshot();
  });

  it("renders a LinkDisabled button", () => {
    const { container } = render(<LinkDisabled />);
    expect(container).toMatchSnapshot();
  });

  it("can be clicked", () => {
    const onClick = vi.fn();
    render(<Button onClick={onClick}>Test</Button>);

    fireEvent.click(screen.getByText("Test"));

    expect(onClick).toHaveBeenCalled();
  });

  it("cannot be clicked when disabled", () => {
    const onClick = vi.fn();
    render(
      <Button onClick={onClick} disabled={true}>
        Test
      </Button>,
    );

    fireEvent.click(screen.getByText("Test"));

    expect(onClick).not.toHaveBeenCalled();
  });

  it("does not submit when disabled", () => {
    const onSubmit = vi.fn();
    render(
      <Button onSubmit={onSubmit} disabled={true}>
        Test
      </Button>,
    );

    fireEvent.click(screen.getByText("Test"));

    expect(onSubmit).not.toHaveBeenCalled();
  });

  it("can be focused when disabled", () => {
    const onClick = vi.fn();
    const onFocus = vi.fn();
    render(
      <Button onClick={onClick} onFocus={onFocus} disabled={true}>
        Test
      </Button>,
    );

    fireEvent.focus(screen.getByText("Test"));

    expect(onFocus).toHaveBeenCalled();
  });
});
