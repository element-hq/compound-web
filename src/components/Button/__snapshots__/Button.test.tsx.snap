// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > accepts an icon 1`] = `
<DocumentFragment>
  <button
    class="_button_2a1efe _hasIcon_2a1efe"
    data-kind="primary"
    data-size="lg"
    role="button"
    tabindex="0"
  >
    <svg
      aria-hidden="true"
      class="_icon_2a1efe"
      fill="currentColor"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m2.975 16.3-1.45 4.95a.936.936 0 0 0 .25 1 .936.936 0 0 0 1 .25l4.95-1.45a10.23 10.23 0 0 0 2.1.712c.717.159 1.45.238 2.2.238a9.738 9.738 0 0 0 3.9-.788 10.098 10.098 0 0 0 3.175-2.137c.9-.9 1.612-1.958 2.137-3.175a9.738 9.738 0 0 0 .788-3.9 9.738 9.738 0 0 0-.787-3.9A10.099 10.099 0 0 0 19.1 4.925c-.9-.9-1.958-1.612-3.175-2.137a9.738 9.738 0 0 0-3.9-.788 9.738 9.738 0 0 0-3.9.788A10.099 10.099 0 0 0 4.95 4.925c-.9.9-1.613 1.958-2.138 3.175a9.738 9.738 0 0 0-.787 3.9 10.179 10.179 0 0 0 .95 4.3Z"
        fill="currentColor"
      />
    </svg>
    With icon
  </button>
</DocumentFragment>
`;

exports[`Button > renders 1`] = `
<DocumentFragment>
  <button
    class="_button_2a1efe"
    data-kind="primary"
    data-size="lg"
    role="button"
    tabindex="0"
  >
    Click me!
  </button>
</DocumentFragment>
`;
